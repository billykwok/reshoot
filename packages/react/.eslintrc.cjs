module.exports = {
  root: false,
  env: {
    browser: true,
    node: true,
    commonjs: true,
    es6: true,
    'jest/globals': true,
  },
  parser: '@babel/eslint-parser',
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',
    ecmaFeatures: { impliedStrict: true, jsx: true },
    babelOptions: { configFile: './babel.config.cjs' },
  },
  settings: {
    react: { version: 'detect' },
    'import/extensions': ['.js', '.jsx', '.json'],
    'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
  },
  plugins: [
    'import',
    'prettier',
    'react',
    'react-hooks',
    'jsx-a11y',
    'testing-library',
    'jest',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended',
    'prettier',
    'plugin:testing-library/react',
    'plugin:jest/recommended',
  ],
  rules: {
    'react/prop-types': 'off',
    'react/jsx-uses-react': 'off',
    'react/react-in-jsx-scope': 'off',
    'jsx-a11y/image-has-content': 'off',
    'jsx-a11y/heading-has-content': 'off',
    'jsx-a11y/anchor-has-content': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'prefer-spread': 'off',
    'consistent-return': 'off',
    'max-len': 'off',
    'no-console': 'off',
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        ecmaVersion: 'latest',
        sourceType: 'module',
        ecmaFeatures: { impliedStrict: true, jsx: true },
        project: ['./tsconfig.json'],
      },
      settings: {
        react: { version: 'detect' },
        'import/extensions': ['.ts', '.tsx'],
        'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
        'import/parsers': { '@typescript-eslint/parser': ['.ts', '.tsx'] },
      },
      plugins: [
        '@typescript-eslint',
        'import',
        'prettier',
        'react',
        'react-hooks',
        'jsx-a11y',
        'testing-library',
        'jest',
      ],
      extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:prettier/recommended',
        'prettier',
        'plugin:testing-library/react',
        'plugin:jest/recommended',
      ],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
        '@typescript-eslint/no-explicit-any': 'off',
        '@typescript-eslint/no-floating-promises': 'off',
        '@typescript-eslint/explicit-function-return-type': 'off',
        '@typescript-eslint/explicit-module-boundary-types': [
          'error',
          { allowArgumentsExplicitlyTypedAsAny: true },
        ],
        '@typescript-eslint/prefer-includes': 'off',
        'react/prop-types': 'off',
        'react/jsx-uses-react': 'off',
        'react/react-in-jsx-scope': 'off',
        'jsx-a11y/image-has-content': 'off',
        'jsx-a11y/heading-has-content': 'off',
        'jsx-a11y/anchor-has-content': 'off',
        'react-hooks/rules-of-hooks': 'error',
      },
    },
  ],
};
