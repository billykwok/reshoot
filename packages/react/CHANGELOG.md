# @reshoot/react

## 2.0.5

### Patch Changes

- a0288f4: fix: support hash character in inline options

## 2.0.4

### Patch Changes

- ec76336: fix: improve rollup setup and update dependencies
- Updated dependencies [ec76336]
  - @reshoot/types@2.0.3

## 2.0.3

### Patch Changes

- 0e54ec5: fix: fix style exports

## 2.0.2

### Patch Changes

- 510a7da: fix: fix incorrect extensions in package.json

## 2.0.1

### Patch Changes

- 226a185: fix: add node version restriction and fix rollup config
- Updated dependencies [226a185]
  - @reshoot/types@2.0.1

## 2.0.0

### Major Changes

- 8d162cb: Introducing Reshoot 2

  Reshoot has been rewritten from ground up to support more use cases. Please refer to README.md for details.

### Patch Changes

- Updated dependencies [8d162cb]
  - @reshoot/types@2.0.0
