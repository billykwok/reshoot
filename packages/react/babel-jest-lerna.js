module.exports = require('babel-jest').createTransformer({
  rootMode: 'upward',
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    'dynamic-import-node',
    ['const-enum', { transform: 'constObject' }],
  ],
});
