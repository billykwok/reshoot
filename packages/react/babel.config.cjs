module.exports = {
  targets: { esmodules: true },
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    ['@babel/preset-typescript', { isTSX: true, allExtensions: true }],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    '@linaria',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      { corejs: { version: 3 }, helpers: true },
    ],
  ],
  compact: true,
  minified: true,
};
