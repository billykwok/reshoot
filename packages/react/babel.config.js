module.exports = {
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    '@babel/preset-react',
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    ['const-enum', { transform: 'constObject' }],
  ],
};
