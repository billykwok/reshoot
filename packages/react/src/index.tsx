import {
  type CSSProperties,
  type ForwardedRef,
  type MutableRefObject,
  type MouseEvent,
  type ReactNode,
  type Ref,
  type RefCallback,
  forwardRef,
  useCallback,
  useRef,
  useState,
} from 'react';
import { css } from '@linaria/core';
import useLifecycleRef from 'use-lifecycle-ref';
import type { ImageMeta } from '@reshoot/types';

import useLazyImage from './useLazyImage';

declare module 'react' {
  interface CSSProperties {
    [index: `--${string}`]: any;
  }
}

const enum State {
  INITIAL,
  FADING,
  LOADED,
  ERROR,
}

const container = css`
  position: relative;
  overflow: hidden;
  background: var(--c);
  &:before {
    content: '';
    display: block;
    width: 100%;
    height: 0;
    padding: 0 0 var(--r);
  }
  > * {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
  > div[aria-hidden='true'] {
    background: var(--c);
    > img {
      width: 100%;
      height: 100%;
      filter: blur(2rem);
    }
  }
`;

const error = css`
  display: flex;
  backdrop-filter: blur(0.5rem);
  align-items: center;
  justify-content: center;
  flex-direction: column;
  color: #f7f7f7;
  text-shadow: #000 0 0 1rem;
  background: rgba(0, 0, 0, 0.5);
  > p {
    margin: 0 0 0.5rem;
  }
  > p:first-child {
    font-size: 1.125rem;
  }
`;

const fadeOut = css`
  animation: fadeOut 0.75s 1 forwards;
  @keyframes fadeOut {
    0% {
      opacity: 1;
    }
    100% {
      opacity: 0;
    }
  }
`;

export type Props = {
  meta: ImageMeta;
  container?:
    | string
    | ((props: {
        ref: Ref<HTMLElement>;
        className: string;
        children: ReactNode;
        style: CSSProperties;
      }) => ReactNode);
  ref?: RefCallback<HTMLElement> | MutableRefObject<HTMLElement>;
  imgRef?: RefCallback<HTMLImageElement> | MutableRefObject<HTMLImageElement>;
  alt?: string;
  className?: string;
  style?: Record<string, any>;
  onClick?: (e: MouseEvent<HTMLElement>) => void;
  [key: string]: any;
};

const Img = forwardRef(function RefReceivingImg(
  {
    meta,
    container: Container = 'div',
    imgRef,
    alt,
    className,
    style,
    ...props
  }: Props,
  ref: ForwardedRef<HTMLElement>
): JSX.Element {
  const { src, srcSet, sizes, aspectRatio, placeholder, color } = meta;
  const [state, setState] = useState(State.INITIAL);
  const instantFade = useRef(false);
  const onLoad = useCallback(
    () => setState(instantFade.current ? State.LOADED : State.FADING),
    []
  );
  const onError = useCallback(() => setState(State.ERROR), []);
  useLazyImage(src, srcSet, sizes, onLoad, onError);
  const _imgRef = useLifecycleRef<HTMLImageElement>({
    onAttach: (element) => {
      if (element.complete) {
        instantFade.current = true;
      }
    },
    ref: imgRef,
  });
  return (
    <Container
      ref={ref}
      className={container + (className ? ' ' + className : '')}
      style={{ '--c': color, '--r': `${aspectRatio}%`, ...style }}
      {...props}
    >
      <img
        ref={_imgRef}
        src={src}
        srcSet={srcSet}
        sizes={sizes}
        alt={alt}
        crossOrigin=""
        aria-busy={State.LOADED !== state && State.ERROR !== state}
      />
      {State.LOADED !== state && (
        <div
          className={
            'js-only' +
            (State.FADING === state && !instantFade.current
              ? ' ' + fadeOut
              : '')
          }
          onAnimationEnd={
            State.FADING === state ? () => setState(State.LOADED) : null
          }
          aria-hidden="true"
        >
          <img src={placeholder} alt="" />
        </div>
      )}
      {State.ERROR === state && (
        <div className={error}>
          <p>Failed to load image</p>
          <p>Please try refreshing the page</p>
        </div>
      )}
    </Container>
  );
});

export { type ImageMeta } from '@reshoot/types';

export default Img;
