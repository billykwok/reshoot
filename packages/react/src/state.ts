export const MANUAL = 0;

export const OFFLINE = 1;

export const ERROR = 2;

export const HIDDEN = 3;

export const LOADING = 4;

export const FADING = 5;

export const LOADED = 6;

export type State =
  | typeof MANUAL
  | typeof OFFLINE
  | typeof ERROR
  | typeof HIDDEN
  | typeof LOADING
  | typeof FADING
  | typeof LOADED;
