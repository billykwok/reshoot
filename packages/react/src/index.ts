import { memo, useRef, useImperativeHandle, forwardRef } from 'react';
import { css } from 'linaria';
import assign from 'object-assign';
import type { RefObject, SyntheticEvent } from 'react';

import h from './h';
import Placeholder from './Placeholder';
import Img from './Img';
import Message from './Message';
import State from './state';
import useImage from './useImage';
import useCx from './useCx';

type Props = {
  src: string;
  alt: string;
  aspectRatio: number;
  className?: string;
  blur?: number;
  color?: string;
  placeholder?: string;
  srcSet?: string;
  target?: string;
  href?: string;
  messages?: {
    [State.MANUAL]: string;
    [State.OFFLINE]: string;
    [State.ERROR]: string;
  };
  onClick?: (e: SyntheticEvent<Element, MouseEvent>) => void;
};

const asContainer = css`
  overflow: hidden;
  position: relative;
  text-decoration: none;
  padding: 0;
  border: none;
  outline: none;
`;

const asButtonContainer = css`
  cursor: pointer;
`;

const Reshoot = (
  {
    src,
    className,
    alt,
    aspectRatio,
    blur = 20,
    color = 'transparent',
    placeholder = null,
    srcSet = null,
    target = '_self',
    href = null,
    messages = {
      MANUAL: 'Not autoloaded in slow network',
      OFFLINE: 'Browser is offline',
      ERROR: 'Fail to load',
    },
    onClick,
    ...rest
  }: Props,
  ref: RefObject<HTMLElement>
) => {
  const innerRef = useRef<HTMLElement>(null);
  useImperativeHandle(ref, () => innerRef.current);
  const [state, setState, download] = useImage(innerRef, src, srcSet);
  const cx = useCx(className);

  const children = [
    h(Placeholder, { color, aspectRatio }),
    h(Img, { color, placeholder, src, srcSet, alt, state, blur }),
    h(Message, { state, text: messages[state] as string }),
  ];

  if (state !== State.INITIAL && state !== State.LOADED) {
    return h(
      'button',
      assign(
        {
          ref: innerRef,
          className: cx(asContainer, asButtonContainer),
          onClick: (e: SyntheticEvent<Element, MouseEvent>) => {
            e.stopPropagation();
            e.preventDefault();
            setState(() => State.INITIAL);
            download();
          },
        },
        rest
      ),
      ...children
    );
  }
  if (href) {
    return h(
      'a',
      assign(
        { ref: innerRef, className: cx(asContainer), target, href, onClick },
        rest
      ),
      ...children
    );
  }
  return h(
    'div',
    assign({ ref: innerRef, className: cx(asContainer) }, rest),
    ...children
  );
};

export default memo(forwardRef(Reshoot));
