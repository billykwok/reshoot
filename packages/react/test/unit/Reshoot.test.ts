import { describe, test, expect } from '@jest/globals';
import { render, act } from '@testing-library/react';
import { createElement, createRef } from 'react';
import { renderToString } from 'react-dom/server';
import * as State from '../../src/state';

describe('Reshoot', () => {
  const download = jest.fn(() => Promise.resolve());
  const baseConfig = {
    src: 'image.jpg',
    alt: 'Test image',
    width: 100,
    height: 80,
  };

  beforeEach(() => {
    jest.doMock(
      '../../src/hooks/useDownload',
      () =>
        ({ __esModule: true, default: () => download } as { __esModule: true })
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('parse valid input', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, { 'data-testid': '123', config: baseConfig, ref })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('customized options for image link', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: {
          ...baseConfig,
          color: '#fff',
          placeholder: 'placeholder.jpg',
        },
        ref,
        target: '_blank',
        href: 'https://example.com',
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('customized options for normal image', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, color: '#fff' },
        ref,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('customized options for error image', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: baseConfig,
        ref,
        _s: State.ERROR,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();

    act(() => dom.click());
    expect(download).toHaveBeenCalledTimes(1);
    expect(dom).toMatchSnapshot();
  });

  test('aspect ratio can be overridden', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const { getByTestId } = render(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, aspectRatio: 2 },
        ref,
        _s: State.ERROR,
      })
    );
    const dom = getByTestId('123');
    expect(dom).toEqual(ref.current);
    expect(dom).toMatchSnapshot();
  });

  test('server side rendering', async () => {
    const { default: Reshoot } = await import('../../src/Reshoot');
    const ref = createRef<HTMLImageElement>();
    const html = renderToString(
      createElement(Reshoot, {
        'data-testid': '123',
        config: { ...baseConfig, aspectRatio: 2 },
        ref,
      })
    );
    expect(html).toMatchSnapshot();
  });
});
