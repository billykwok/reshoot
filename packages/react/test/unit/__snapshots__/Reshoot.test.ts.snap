// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reshoot aspect ratio can be overridden 1`] = `
<button
  class="ax4jui0"
  data-testid="123"
  style="--r: 0.5;"
>
  <div>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <div />
  </div>
</button>
`;

exports[`Reshoot customized options for error image 1`] = `
<button
  class="ax4jui0"
  data-testid="123"
  style="--r: 0.8;"
>
  <div>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <div />
  </div>
</button>
`;

exports[`Reshoot customized options for error image 2`] = `
<button
  class="ax4jui0"
  data-testid="123"
  style="--r: 0.8;"
>
  <div>
    <svg
      viewBox="0 0 100 100"
    >
      <path
        d="M79.4 56a30 30 0 1 1-1.7-17.5m2-19v20h-20"
      />
    </svg>
    <div />
  </div>
</button>
`;

exports[`Reshoot customized options for image link 1`] = `
<a
  class="ax4jui0"
  data-testid="123"
  href="https://example.com"
  style="color: rgb(255, 255, 255); --r: 0.8;"
  target="_blank"
>
  <img
    alt="Test image"
    class="a1rj2wdy"
    decoding="async"
    height="80"
    loading="lazy"
    src="placeholder.jpg"
    style="color: rgb(255, 255, 255);"
    width="100"
  />
</a>
`;

exports[`Reshoot customized options for normal image 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="color: rgb(255, 255, 255); --r: 0.8;"
>
  <img
    alt="Test image"
    class="acf3p1m"
    decoding="async"
    height="80"
    loading="lazy"
    src="image.jpg"
    style="color: rgb(255, 255, 255);"
    width="100"
  />
</div>
`;

exports[`Reshoot parse valid input 1`] = `
<div
  class="ax4jui0"
  data-testid="123"
  style="--r: 0.8;"
>
  <img
    alt="Test image"
    class="acf3p1m"
    decoding="async"
    height="80"
    loading="lazy"
    src="image.jpg"
    width="100"
  />
</div>
`;

exports[`Reshoot server side rendering 1`] = `"<div class=\\"ax4jui0\\" style=\\"--r:0.5\\" data-testid=\\"123\\"><img class=\\"acf3p1m\\" loading=\\"lazy\\" decoding=\\"async\\" src=\\"image.jpg\\" width=\\"100\\" height=\\"80\\" alt=\\"Test image\\"/></div>"`;
